### PR TITLE
BUG/TST: Fix TimedeltaIndex comparisons with invalid types

### LINKED ISSUE #23063: BUG: TimedeltaIndex == integer raises TypeError
```
pd.TimedeltaIndex(['1D']) == 1
[...]
TypeError: cannot compare a TimedeltaIndex with type int

pd.TimedeltaIndex(['1D']) == 1.
[...]
TypeError: cannot compare a TimedeltaIndex with type float
```

### PR DESCRIPTION
**Pull request by jbrockmendel**
Parametrize a handful of tests.

- [x] closes #23063
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/arrays/datetimes.py [hunk 1]
```diff
@@ -217,6 +220,12 @@ def __new__(cls, values, freq=None, tz=None, dtype=None):
         # if dtype has an embedded tz, capture it
         tz = dtl.validate_tz_from_dtype(dtype, tz)
 
+        if not hasattr(values, "dtype"):
+            if np.ndim(values) == 0:
+                # i.e. iterator
+                values = list(values)
```
Comment (jreback) [inline pandas/core/arrays/datetimes.py]: you have a test hit this?

File: pandas/core/ops.py [hunk 2]
```diff
@@ -1115,7 +1115,7 @@ def dispatch_to_series(left, right, func, str_rep=None, axis=None):
     import pandas.core.computation.expressions as expressions
 
     right = lib.item_from_zerodim(right)
-    if lib.is_scalar(right):
+    if lib.is_scalar(right) or np.ndim(right) == 0:
```
Comment (jreback) [inline pandas/core/ops.py]: why is this needed? this *should* already be handled by is_scalar

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/datetimes.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: Yes

File: pandas/core/ops.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/ops.py]: because there are lots of weird things a user _could_ pass that are zero-dim but not caught by is_scalar.  The test case that hits this just passes `object()`

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!
